derp/derphttp/derphttp_client.go

```diff
@@ -243,6 +292,133 @@ func (c *Client) connect(ctx context.Context, caller string) (client *derp.Clien
 	return c.client, nil
 }
 
+func (c *Client) dialURL(ctx context.Context) (net.Conn, error) {
+	host := c.url.Hostname()
+	hostOrIP := host
+
+	var stdDialer dialer = new(net.Dialer)
+	var dialer = stdDialer
+	if wrapDialer != nil {
+		dialer = wrapDialer(dialer)
+	}
+
+	if c.DNSCache != nil {
+		ip, err := c.DNSCache.LookupIP(ctx, host)
+		if err == nil {
+			hostOrIP = ip.String()
+		}
+		if err != nil && dialer == stdDialer {
+			// Return an error if we're not using a dial
+			// proxy that can do DNS lookups for us.
+			return nil, err
+		}
+	}
+
+	tcpConn, err := dialer.DialContext(ctx, "tcp", net.JoinHostPort(hostOrIP, urlPort(c.url)))
+	if err != nil {
+		return nil, fmt.Errorf("dial of %v: %v", host, err)
+	}
+	return tcpConn, nil
+}
+
+func (c *Client) dialRegion(ctx context.Context, reg *tailcfg.DERPRegion) (net.Conn, *tailcfg.DERPNode, error) {
+	if len(reg.Nodes) == 0 {
+		return nil, nil, fmt.Errorf("no nodes for %s", c.targetString(reg))
+	}
+	var firstErr error
+	for _, n := range reg.Nodes {
+		if n.STUNOnly {
+			continue
+		}
+		c, err := c.dialNode(ctx, n)
+		if err == nil {
+			return c, n, nil
+		}
+		if firstErr == nil {
+			firstErr = err
+		}
+	}
+	return nil, nil, firstErr
+}
+
+func (c *Client) dialContext(ctx context.Context, proto, addr string) (net.Conn, error) {
+	var stdDialer dialer = new(net.Dialer)
+	var dialer = stdDialer
+	if wrapDialer != nil {
+		dialer = wrapDialer(dialer)
+	}
+	return dialer.DialContext(ctx, proto, addr)
+}
+
+func shouldDialProto(s string, pred func(netaddr.IP) bool) bool {
+	if s == "" {
+		return true
+	}
+	ip, _ := netaddr.ParseIP(s)
+	return pred(ip)
+}
+
+func (c *Client) dialNode(ctx context.Context, n *tailcfg.DERPNode) (net.Conn, error) {
+	type res struct {
+		c   net.Conn
+		err error
+	}
+	resc := make(chan res) // must be unbuffered
+	ctx, cancel := context.WithTimeout(ctx, 1500*time.Millisecond)
```

Comment:
So nodes for a region are dialed in order, with a 1.5s timeout before trying the next node? Load balancing comes from the server in the ordering of the Nodes slice?

If so maybe put that in a comment in dialRegion.

Reply:
Done.